### ISSUE TITLE
Improve parallelization of test suites

### ISSUE THREAD

**Issue by sksamuel** [kotest/kotest#1760]
At the moment we allow users to set a number of threads to be used as a pool to execute specs. We support something similar for tests.

Instead we could allow two modes for spec execution - purely sequential and parallel (with the opt-out for certain specs to always run in isolation like now).

Then instead of asking users to specify the number of threads, they can provide a coroutine dispatcher, which is used to launch a coroutine per spec. This would allow greater throughput when using suspending functions, as they would suspend, freeing up the dispatcher for other coroutines. At the moment, if they suspend, they just block the thread since all specs ultimately fall back to a runBlocking block at the java executor level.

This could be carried through into test cases themselves, each root test case can be dispatched into a coroutine, allowing these to run in parallel in the same way.

Thus:
 - a setting for sequential / parallel spec submission
 - a setting for test case parallel execution
 - a dispatcher to control the concurrency

**Comment by jschneidereit** [comment]
This looks fantastic and sounds like just what I need, thanks for opening this!

**Comment by sksamuel** [comment]
Some notes for an implementor.

`submitBatch` and `submitAll` methods in `KotestEngine` are where the spec level execution pool would need to be replaced with coroutines that are launched onto a user specified (or default fallback) dispatcher.

**Comment by sksamuel** [comment]
This is implemented in master now.

There is a `ConcurrencyMode` configuration setting that allows the user to select how coroutines are created for tests.
Non public documentation is here: https://github.com/kotest/kotest/blob/master/docs/concurrency.md

**Comment by sksamuel** [comment]
This has been updated in master, documentation to follow once RC1 is released.

**Comment by jkasten2** [comment]
> This is implemented in master now.
> 
> There is a `ConcurrencyMode` configuration setting that allows the user to select how coroutines are created for tests. Non public documentation is here: https://github.com/kotest/kotest/blob/master/docs/concurrency.md

Looks like the documentation is public here:
https://github.com/kotest/kotest/blob/master/documentation/docs/framework/concurrency.md


> This has been updated in master, documentation to follow once RC1 is released.

Don't see a "Concurrency" section on the page. (would expect it since it is in the docs/framework directory in github)
https://kotest.io/docs/framework/framework.html

Was "ConcurrencyMode" dropped and become something else?
I tired using "kotest.framework.concurrency.mode" noted in the [docs/framework/concurrency.md](https://github.com/kotest/kotest/blob/v5.2.2/documentation/docs/framework/concurrency.md#concurrency-mode) for v5.2.2 but it seems to have no effect and don't see as a constant at all in the source of this repo.

**Comment by sksamuel** [comment]
@jkasten2 I think this is more up to date https://kotest.io/docs/framework/concurrency.html

**Comment by jkasten2** [comment]
> @jkasten2 I think this is more up to date https://kotest.io/docs/framework/concurrency.html

@sksamuel Thanks for the quick response, it looks like https://kotest.io/docs/framework/concurrency.html is just the same as https://github.com/kotest/kotest/blob/master/documentation/docs/framework/concurrency.md (as probably expected). It notes `"kotest.framework.concurrency.mode"` for this setting as well, but it doesn't have any effect. Also don't see the `concurrencyMode` option anywhere either.

Seems like there is a problem with the docs or with kotest itself since `"kotest.framework.concurrency.mode"` doesn't seem to have any effect. I'll create a new issue, since this one is focused on a feature request instead of a bug.
